// cdk8s-cli/lib/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelmChartApiVersion = exports.SynthesisFormat = void 0;
exports.readConfigSync = readConfigSync;
exports.addImportToConfig = addImportToConfig;
const fs = __importStar(require("fs-extra"));
const yaml = __importStar(require("yaml"));
const CONFIG_FILE = 'cdk8s.yaml';
var SynthesisFormat;
(function (SynthesisFormat) {
    SynthesisFormat["PLAIN"] = "plain";
    SynthesisFormat["HELM"] = "helm";
})(SynthesisFormat || (exports.SynthesisFormat = SynthesisFormat = {}));
var HelmChartApiVersion;
(function (HelmChartApiVersion) {
    HelmChartApiVersion["V1"] = "v1";
    HelmChartApiVersion["V2"] = "v2";
})(HelmChartApiVersion || (exports.HelmChartApiVersion = HelmChartApiVersion = {}));
function readConfigSync() {
    if (fs.existsSync(CONFIG_FILE)) {
        const config = {
            ...yaml.parse(fs.readFileSync(CONFIG_FILE, 'utf-8')),
        };
        return config;
    }
    return undefined;
}
async function addImportToConfig(source) {
    if (!fs.existsSync(CONFIG_FILE)) {
        // cdk8s import might be executed outside of an application (i.e without a config file)
        return;
    }
    let curConfig = yaml.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
    const curImports = curConfig.imports ?? [];
    if (!curImports.includes(source)) {
        const importsList = curConfig.imports ?? [];
        importsList.push(source);
        let config = {
            ...curConfig,
            imports: importsList,
        };
        await fs.outputFile(CONFIG_FILE, yaml.stringify(config));
    }
}
