// jsii-rosetta/lib/translate.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Translator_diagnostics;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchSnippetTranslator = exports.SnippetTranslator = exports.Translator = void 0;
exports.translateTypeScript = translateTypeScript;
exports.makeRosettaDiagnostic = makeRosettaDiagnostic;
exports.makeTimingDiagnostic = makeTimingDiagnostic;
exports.extractTimingInfo = extractTimingInfo;
exports.formatTimingTable = formatTimingTable;
exports.rosettaDiagFromTypescript = rosettaDiagFromTypescript;
const path = require("node:path");
const node_util_1 = require("node:util");
const ts = require("typescript");
const languages_1 = require("./languages");
const record_references_1 = require("./languages/record-references");
const target_language_1 = require("./languages/target-language");
const logging = require("./logging");
const o_tree_1 = require("./o-tree");
const renderer_1 = require("./renderer");
const snippet_1 = require("./snippet");
const submodule_reference_1 = require("./submodule-reference");
const schema_1 = require("./tablets/schema");
const tablets_1 = require("./tablets/tablets");
const syntax_kind_counter_1 = require("./typescript/syntax-kind-counter");
const ts_compiler_1 = require("./typescript/ts-compiler");
const visible_spans_1 = require("./typescript/visible-spans");
const util_1 = require("./util");
function translateTypeScript(source, visitor, options = {}) {
    const translator = new SnippetTranslator({ visibleSource: source.contents, location: { api: { api: 'file', fileName: source.fileName } } }, options);
    const translated = translator.renderUsing(visitor);
    return {
        translation: translated,
        diagnostics: translator.diagnostics.map(rosettaDiagFromTypescript),
    };
}
/**
 * Translate one or more TypeScript snippets into other languages
 *
 * Can be configured to fully typecheck the samples, or perform only syntactical
 * translation.
 */
class Translator {
    get diagnostics() {
        return ts.sortAndDeduplicateDiagnostics(__classPrivateFieldGet(this, _Translator_diagnostics, "f")).map(rosettaDiagFromTypescript);
    }
    constructor(includeCompilerDiagnostics) {
        this.includeCompilerDiagnostics = includeCompilerDiagnostics;
        this.compiler = new ts_compiler_1.TypeScriptCompiler();
        _Translator_diagnostics.set(this, []);
    }
    /**
     * Return the snippet translator for the given snippet
     *
     * We used to cache these, but each translator holds on to quite a bit of memory,
     * so we don't do that anymore.
     */
    translatorFor(snippet) {
        const translator = new SnippetTranslator(snippet, {
            compiler: this.compiler,
            includeCompilerDiagnostics: this.includeCompilerDiagnostics,
        });
        return translator;
    }
    /**
     * Translates a single snippet in its own TS context.
     */
    translate(snip, languages = Object.values(languages_1.TargetLanguage)) {
        const translator = this.translatorFor(snip);
        const translated = this.translateSnippet(snip, translator, languages);
        return translated;
    }
    /**
     * Translates a batch of snippets, using a shared TS context.
     */
    translateSnippets(snippets, languages = Object.values(languages_1.TargetLanguage)) {
        const start = performance.now();
        logging.debug(`Translating batch of ${snippets.length} snippets`);
        const res = this.translateBatch(snippets, languages);
        const duration = performance.now() - start;
        logging.debug(`Completed batch ${(0, node_util_1.inspect)({
            duration: `${(duration / 1000).toFixed(2)}s`,
        })}`);
        return res;
    }
    translateBatch(snippets, languages = Object.values(languages_1.TargetLanguage)) {
        const translatedSnippets = [];
        const batchTranslator = new BatchSnippetTranslator(snippets, {
            compiler: this.compiler,
            includeCompilerDiagnostics: this.includeCompilerDiagnostics,
        });
        for (const [snippet, translator] of batchTranslator) {
            translatedSnippets.push(this.translateSnippet(snippet, translator, languages));
        }
        return translatedSnippets;
    }
    translateSnippet(snippet, translator, languages) {
        const translations = (0, util_1.mkDict)(languages.flatMap((lang, idx, array) => {
            if (array.slice(0, idx).includes(lang)) {
                return [];
            }
            const languageConverterFactory = languages_1.TARGET_LANGUAGES[lang];
            const translated = translator.renderUsing(languageConverterFactory.createVisitor());
            return [[lang, { source: translated, version: languageConverterFactory.version }]];
        }));
        if (snippet.parameters?.infused === undefined) {
            __classPrivateFieldGet(this, _Translator_diagnostics, "f").push(...translator.diagnostics);
        }
        return tablets_1.TranslatedSnippet.fromSchema({
            translations: {
                ...translations,
                [schema_1.ORIGINAL_SNIPPET_KEY]: { source: snippet.visibleSource, version: '0' },
            },
            location: snippet.location,
            didCompile: translator.didSuccessfullyCompile,
            fqnsReferenced: translator.fqnsReferenced(),
            fullSource: (0, snippet_1.completeSource)(snippet),
            syntaxKindCounter: translator.syntaxKindCounter(),
        });
    }
}
exports.Translator = Translator;
_Translator_diagnostics = new WeakMap();
/**
 * Internal implementation of a single TypeScript snippet translator.
 *
 * Consumers should either use `SnippetTranslator` or `BatchSnippetTranslator`.
 */
class InternalSnippetTranslator {
    constructor(snippet, compilation, options) {
        this.compilation = compilation;
        this.options = options;
        this.translateDiagnostics = [];
        this.compileDiagnostics = [];
        // Respect '/// !hide' and '/// !show' directives
        // Use the actual compiled source text to ensure spans match the AST
        this.visibleSpans = visible_spans_1.Spans.visibleSpansFromSource(compilation.rootFile.text);
        // Find submodule references on explicit imports
        this.submoduleReferences = submodule_reference_1.SubmoduleReference.inSourceFile(compilation.rootFile, this.compilation.program.getTypeChecker());
        // This makes it about 5x slower, so only do it on demand
        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
        this.tryCompile = (this.options.includeCompilerDiagnostics || snippet.strict) ?? false;
        if (this.tryCompile) {
            const program = this.compilation.program;
            const diagnostics = [
                ...neverThrowing(program.getGlobalDiagnostics)(),
                ...neverThrowing(program.getSyntacticDiagnostics)(this.compilation.rootFile),
                ...neverThrowing(program.getDeclarationDiagnostics)(this.compilation.rootFile),
                ...neverThrowing(program.getSemanticDiagnostics)(this.compilation.rootFile),
            ];
            if (snippet.strict) {
                // In a strict assembly, so we'll need to brand all diagnostics here...
                for (const diag of diagnostics) {
                    (0, util_1.annotateStrictDiagnostic)(diag);
                }
            }
            this.compileDiagnostics.push(...diagnostics);
        }
    }
    /**
     * Returns a boolean if compilation was attempted, and undefined if it was not.
     */
    get didSuccessfullyCompile() {
        return this.tryCompile ? this.compileDiagnostics.length === 0 : undefined;
    }
    renderUsing(visitor) {
        const converter = new renderer_1.AstRenderer(this.compilation.rootFile, this.compilation.program.getTypeChecker(), visitor, this.options, 
        // If we support transitive submodule access, don't provide a submodule reference map.
        (0, target_language_1.supportsTransitiveSubmoduleAccess)(visitor.language) ? undefined : this.submoduleReferences);
        const converted = converter.convert(this.compilation.rootFile);
        this.translateDiagnostics.push(...filterVisibleDiagnostics(converter.diagnostics, this.visibleSpans));
        return (0, o_tree_1.renderTree)(converted, { indentChar: visitor.indentChar, visibleSpans: this.visibleSpans });
    }
    syntaxKindCounter() {
        const kindCounter = new syntax_kind_counter_1.SyntaxKindCounter(this.visibleSpans);
        return kindCounter.countKinds(this.compilation.rootFile);
    }
    fqnsReferenced() {
        const visitor = new record_references_1.RecordReferencesVisitor(this.visibleSpans);
        const converter = new renderer_1.AstRenderer(this.compilation.rootFile, this.compilation.program.getTypeChecker(), visitor, this.options, this.submoduleReferences);
        converter.convert(this.compilation.rootFile);
        return visitor.fqnsReferenced();
    }
    get diagnostics() {
        return ts.sortAndDeduplicateDiagnostics(this.compileDiagnostics.concat(this.translateDiagnostics));
    }
}
/**
 * Translate a single TypeScript snippet
 */
class SnippetTranslator extends InternalSnippetTranslator {
    constructor(snippet, options = {}) {
        const compiler = options.compiler ?? new ts_compiler_1.TypeScriptCompiler();
        const source = (0, snippet_1.completeSource)(snippet);
        const fakeCurrentDirectory = snippet.parameters?.[snippet_1.SnippetParameters.$COMPILATION_DIRECTORY] ?? process.cwd();
        const compilation = compiler.compileBatchInMemory([
            {
                filename: removeSlashes((0, snippet_1.formatLocation)(snippet.location)),
                contents: source,
            },
        ], fakeCurrentDirectory);
        super(snippet, {
            program: compilation.program,
            rootFile: compilation.rootFiles[0],
        }, options);
    }
}
exports.SnippetTranslator = SnippetTranslator;
/**
 * Translate a single TypeScript snippet
 */
class BatchSnippetTranslator {
    constructor(snippets, options = {}) {
        this.snippets = snippets;
        this.options = options;
        const compiler = options.compiler ?? new ts_compiler_1.TypeScriptCompiler();
        const workingDir = options.compilationDirectory ?? process.cwd();
        const sources = snippets.map((snippet) => {
            const snippetLoc = snippet.parameters?.[snippet_1.SnippetParameters.$COMPILATION_DIRECTORY];
            const filename = removeSlashes((0, snippet_1.formatLocation)(snippet.location));
            return {
                filename: snippetLoc ? path.relative(workingDir, path.join(snippetLoc, filename)) : filename,
                contents: (0, snippet_1.completeSource)(snippet),
            };
        });
        this.compilation = compiler.compileBatchInMemory(sources, workingDir);
    }
    *[Symbol.iterator]() {
        for (const [idx, snippet] of this.snippets.entries()) {
            const rootFile = this.compilation.rootFiles[idx];
            const translator = new InternalSnippetTranslator(snippet, {
                program: this.compilation.program,
                rootFile,
            }, this.options);
            yield [snippet, translator];
        }
    }
}
exports.BatchSnippetTranslator = BatchSnippetTranslator;
/**
 * Intercepts all exceptions thrown by the wrapped call, and logs them to
 * console.error instead of re-throwing, then returns an empty array. This
 * is here to avoid compiler crashes due to broken code examples that cause
 * the TypeScript compiler to hit a "Debug Failure".
 */
function neverThrowing(call) {
    return (...args) => {
        try {
            return call(...args);
        }
        catch (err) {
            const isExpectedTypescriptError = err.message.includes('Debug Failure');
            if (!isExpectedTypescriptError) {
                console.error(`Failed to execute ${call.name}: ${err}`);
            }
            return [];
        }
    };
}
/**
 * Hide diagnostics that are rosetta-sourced if they are reported against a non-visible span
 */
function filterVisibleDiagnostics(diags, visibleSpans) {
    return diags.filter((d) => d.source !== 'rosetta' || d.start === undefined || visibleSpans.containsPosition(d.start));
}
function makeRosettaDiagnostic(isError, formattedMessage) {
    return { isError, formattedMessage, isFromStrictAssembly: false };
}
function makeTimingDiagnostic(snippetKey, name, durationMs) {
    return {
        isError: false,
        isFromStrictAssembly: false,
        formattedMessage: '',
        timingInfo: { snippetKey, name, durationMs },
    };
}
function extractTimingInfo(diagnostics) {
    const timings = [];
    const regular = [];
    for (const diag of diagnostics) {
        if (diag.timingInfo) {
            timings.push(diag.timingInfo);
        }
        else {
            regular.push(diag);
        }
    }
    return { timings, diagnostics: regular };
}
function formatTimingTable(timings) {
    if (timings.length === 0) {
        return '';
    }
    const totalTime = timings.reduce((sum, t) => sum + t.durationMs, 0);
    const sorted = timings.sort((a, b) => b.durationMs - a.durationMs).slice(0, 10);
    const lines = [
        '',
        '=== Top 10 Slowest Snippets ===',
        'Rank | Time (s) | % of Total | Snippet ',
        '-----|----------|------------|--------------------------------',
    ];
    for (const [idx, timing] of sorted.entries()) {
        const timeS = (timing.durationMs / 1000).toFixed(2).padStart(8);
        const pct = ((timing.durationMs / totalTime) * 100).toFixed(1).padStart(9) + '%';
        lines.push(`${(idx + 1).toString().padEnd(4)} | ${timeS} | ${pct} | ${timing.name}`);
    }
    lines.push('');
    lines.push(`Total translation time: ${(totalTime / 1000).toFixed(2)}s`);
    lines.push('');
    return lines.join('\n');
}
/**
 * Turn TypeScript diagnostics into Rosetta diagnostics
 */
function rosettaDiagFromTypescript(diag) {
    return {
        isError: diag.category === ts.DiagnosticCategory.Error,
        isFromStrictAssembly: (0, util_1.hasStrictBranding)(diag),
        formattedMessage: ts.formatDiagnosticsWithColorAndContext([diag], DIAG_HOST),
    };
}
const DIAG_HOST = {
    getCurrentDirectory() {
        return '.';
    },
    getCanonicalFileName(fileName) {
        return fileName;
    },
    getNewLine() {
        return '\n';
    },
};
/**
 * Remove slashes from a "where" description, as the TS compiler will interpret it as a directory
 * and we can't have that for compiling literate files
 */
function removeSlashes(x) {
    return x.replace(/\/|\\/g, '.');
}
