// jsii-rosetta/lib/translate_all_worker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.singleThreadedTranslateAll = singleThreadedTranslateAll;
/**
 * Pool worker for extract.ts
 */
const node_util_1 = require("node:util");
const workerpool = require("workerpool");
const logging = require("./logging");
const snippet_1 = require("./snippet");
const key_1 = require("./tablets/key");
const translate_1 = require("./translate");
function translateBatch(request) {
    // because we are in a worker process we need to explicitly configure the log level again
    logging.configure({ level: request.logLevel ?? logging.Level.QUIET, prefix: request.workerName });
    if (process.env.TIMING === '1' && request.batchSize) {
        logging.warn('TIMING=1 is not supported in batch compilation mode');
    }
    const result = request.batchSize
        ? batchTranslateAll(request.snippets, request.includeCompilerDiagnostics)
        : singleThreadedTranslateAll(request.snippets, request.includeCompilerDiagnostics);
    return {
        translatedSchemas: result.translatedSnippets.map((s) => s.snippet),
        diagnostics: result.diagnostics,
    };
}
function batchTranslateAll(snippets, includeCompilerDiagnostics) {
    const translatedSnippets = new Array();
    const failures = new Array();
    const translator = new translate_1.Translator(includeCompilerDiagnostics);
    try {
        const results = translator.translateSnippets(snippets);
        translatedSnippets.push(...results);
    }
    catch (e) {
        failures.push((0, translate_1.makeRosettaDiagnostic)(true, `rosetta: error translating batch: ${e}\n${e.stack}`));
    }
    return {
        translatedSnippets,
        diagnostics: [...translator.diagnostics, ...failures],
    };
}
/**
 * Translate the given snippets using a single compiler
 */
function singleThreadedTranslateAll(snippets, includeCompilerDiagnostics) {
    const translatedSnippets = new Array();
    const failures = new Array();
    const timings = new Array();
    const translator = new translate_1.Translator(includeCompilerDiagnostics);
    for (const block of snippets) {
        const start = performance.now();
        const currentSnippetKey = (0, key_1.snippetKey)(block);
        logging.debug(`Translating ${currentSnippetKey} ${(0, node_util_1.inspect)(block.parameters ?? {})}`);
        try {
            translatedSnippets.push(translator.translate(block));
        }
        catch (e) {
            failures.push((0, translate_1.makeRosettaDiagnostic)(true, `rosetta: error translating snippet: ${e}\n${e.stack}\n${block.completeSource}`));
        }
        const timing = (0, translate_1.makeTimingDiagnostic)(currentSnippetKey, (0, snippet_1.formatLocation)(block.location), performance.now() - start);
        timings.push(timing);
        logging.debug(`Completed ${timing.timingInfo.snippetKey} ${(0, node_util_1.inspect)({
            duration: `${(timing.timingInfo.durationMs / 1000).toFixed(2)}s`,
        })}`);
    }
    return {
        translatedSnippets,
        diagnostics: [...translator.diagnostics, ...failures, ...timings],
    };
}
workerpool.worker({ translateBatch });
